### ISSUE TITLE
Color field: migration for content with mixed formats

### ISSUE THREAD

**Issue by mynameisfreedom** [k-next/kirby#6784]
Hi team, as always your new features are great, and it is fantastic that the [Color field](https://getkirby.com/docs/reference/panel/fields/color) is in the core. 

## Description
But I have one issue with migration from [hananils/kirby-colors](https://github.com/hananils/kirby-colors) plugin to the new Color field. Migration to the Color field is smooth except for one thing: hananils/kirby-colors plugin used rgba color when alpha is set to true. So, at the moment it is not possible to migrate those to the Kirby Color field. 

hananils/kirby-colors saved alpha: true fields in this format rgba(0,0,0,0.3) so when I migrate to new Color field there is an error for every field that used opacity. And because every Layout Builder section in my Layout Builder has a color opacity overlay option for the section background image, that field is used a lot. 🫣

## Expected behavior
If you could either:
- add a script to change rgba inputs into new rgb format (example: rgba(0,0,0,0.3) into rgb(0 0 0 /0.3)) on save
- or add support for legacy rgba colors

hananils/kirby-colors still work, but I doubt that will be the case for too much longer. 

Maybe this is a small thing, but it would help hundreds of websites migrate smoothly to the new Color field. 

Thank you in advance. 🙏

**Comment by distantnative** [comment]
I thought this would already work. Have you the `alpha: true` option set? https://getkirby.com/docs/reference/panel/fields/color#alpha

**Comment by mynameisfreedom** [comment]
Yes, I have `alpha: true` option set but it still gives me an error during saving in the panel. 

Maybe your script don't support rgba(0,0,0,0.3) but support rgba(0, 0, 0, 0.3). I will see to test this.

**Comment by mynameisfreedom** [comment]
Hmm, now in my test everything went smoothly... It migrated the color properly. And a few hours ago it gave me an error on every try. 

You can feely close this, I will go for further migration.

**Comment by mynameisfreedom** [comment]
Ah, on further migration I do get an error. Please leave this open, I need to explore more on why this is happening

![Screenshot 11-09-2024 15 03 42](https://github.com/user-attachments/assets/7e212b0c-6e43-492d-98fa-581e3581d5de)

**Comment by mynameisfreedom** [comment]
Sorry for the delay. I think I found what is making an error. 

The new [Kirby color field](https://getkirby.com/docs/reference/panel/fields/color#field-properties) only supports one format at a time (hex or RGB or HSL), but old [Kirby colors plugin](https://github.com/hananils/kirby-colors) supported all formats at once. 

![screenshot-localhost-2024 11 18-20_15_12](https://github.com/user-attachments/assets/574a36d0-7f5b-4738-9304-22a2ec836adc)

For the same field in the Layout section on one page admins often used hex colors if the color is not transparent and RGB if the color is transparent. In that case, Kirby's color field drops an error for a color format it doesn't support. So, to make the migration, the admin would need to go from field to field and manually change color to get the supported format. 

I reverted to the old [Kirby colors plugin](https://github.com/hananils/kirby-colors) fields all fields with `alpha: true`, for the time being. 

Could you do something about this? 
Can you make a new Kirby Color field, for example: although it has `format: hex` and the old field value is RGB, don't report error, but convert the RGB value to hex? Or just don't throw an error?

**Comment by distantnative** [comment]
@mynameisfreedom I can't reproduce this still:

In the blueprint I have

```yaml
hexFormat:
  type: color
  format: hex
  alpha: true
```

In the content file:

```
HexFormat: rgba(230, 70, 30, 0.7)
```

Panel loads like this:
<img width="311" alt="Screenshot 2024-12-07 at 19 30 36" src="https://github.com/user-attachments/assets/cbb128be-5650-401b-8743-688bd26a57d0">
(working with the RGBA value even though the storing format is set to HEX)

If I focus the input now and click outside again, it'll convert it to the equivalent HEX code:
<img width="206" alt="Screenshot 2024-12-07 at 19 31 23" src="https://github.com/user-attachments/assets/2c87d6d9-145a-48cf-855d-064d63e83181">

**Comment by distantnative** [comment]
Closing this for now. Please let me know if you have additional information how to reproduce it.

**Comment by mynameisfreedom** [comment]
Hi @distantnative, sorry for the delay, I couldn't find time earlier. 

I recorded a short video of the issue, please check it out https://streamable.com/zkj6kq

Testing this on the Kirby 4.6.1 and PHP 8.3.16

**Comment by distantnative** [comment]
@mynameisfreedom Thanks for the video, that made it clear how you are using it and what the problem is occurring. Differently to my own tests, you are not touching the field itself after switching t the new color field, but just trying to save the page. If you would focus the field and click away or save, the value would automatically be converted to the current format (`hex` as the default as you have not entered any different format in the blueprint). But since you are not interacting with the field at all, the field cannot make this automatic conversion but will throw an error as the value arrives on the backend in a different format than the current one (`hex`).

What I would suggest to do:
1. Change the format in the blueprint to `format: rgb`
2. This should allow saving your current RGB-formatted values again
3. When any of your editors pastes HEX codes into the field, those will be converted to matching RGB colors right on the spot.

For mixed content, I have no good answer right now as Kirby doesn't have a good way here yet to recognize that the value remains unchanged (and in that case allow it/skip the format validation).

**Comment by mynameisfreedom** [comment]
Thank you very much @distantnative for your effort. 

The biggest difference between https://github.com/hananils/kirby-colors field and Kirby Color field is that old Colors plugin didn't have option to disallow some color types, and new Kirby Color field does. 

So unfortunately, I'm bit of stuck here. But what can we do, we can't fix everything.

**Comment by distantnative** [comment]
I know, but I think the validation of the color format is actually a good thing for the core field. 

You could consider overwriting the PHP part of the core field with basically the core field code but remove the validation part.

**Comment by mynameisfreedom** [comment]
I agree with you, and I didn't like the lack of color format restriction in the old Kirby Colors plugin. 

If it were for the ordinary website, I would migrate without too much concern, it is easily fixable, but because it is Zero One with hundreds of websites, I wanted to avoid users getting errors when saving after the next update. 

I can only hope that not too many of them changed the default RGB format to something else. 

But I am going for migration, can't keep the old Colors field anymore, I will write a clear warning for this update about that, and help users with it. 

Thanks!

**Comment by mynameisfreedom** [comment]
I will close this issue. Again, many thanks for the effort.
